src/clang-cpp-frontend/clang_cpp_convert.cpp

```diff
@@ -737,9 +741,15 @@ bool clang_cpp_convertert::get_constructor_call(
 
   // Try to get the object that this constructor is constructing
   auto it = ASTContext->getParents(constructor_call).begin();
-
   const clang::Decl *objectDecl = it->get<clang::Decl>();
-  if(!objectDecl)
+
+  /*
+   * Distinguish `Foo foo = new foo();` from `Foo foo = foo();`
+   * The latter involves a temporary object and copy constructor call
+   * where the the temporary object construction is considered an argument to
+   * the cpy ctor call's AST node. We shouldn't make a new_object in this case.
+   */
+  if(!objectDecl && constructor_call.arguments().empty())
```

Comment:
what happens when the object's constructor has one argument but we construct it using `new`?

Reply:
now fixed